File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/DefinitionArgument.scala
package org.enso.compiler.core.ir

import org.enso.compiler.core.IR
import org.enso.compiler.core.IR.{randomId, Identifier, ToStringHelper}

/** Definition-site arguments in Enso. */
sealed trait DefinitionArgument extends IR {

  /** The name of the argument. */
  val name: Name

  /** The type of the argument */
  val ascribedType: Option[Expression]

  /** The default value of the argument. */
  val defaultValue: Option[Expression]

  /** Whether or not the argument is suspended. */
  val suspended: Boolean

  /** @inheritdoc */
  override def mapExpressions(
    fn: Expression => Expression
  ): DefinitionArgument

  /** @inheritdoc */
  override def setLocation(
    location: Option[IdentifiedLocation]
  ): DefinitionArgument

  /** @inheritdoc */
  override def duplicate(
    keepLocations: Boolean   = true,
    keepMetadata: Boolean    = true,
    keepDiagnostics: Boolean = true,
    keepIdentifiers: Boolean = false
  ): DefinitionArgument

  def withName(ir: Name): DefinitionArgument
}

object DefinitionArgument {

  /** The representation of an argument from a [[Function]] or
    * [[module.scope.Definition.Data]] definition site.
    *
    * To create an ignored argument, the argument name should be an
    * [[Name.Blank]].
    *
    * @param name         the name of the argument
    * @param ascribedType the explicitly ascribed type of the argument, if
    *                     present
    * @param defaultValue the default value of the argument, if present
    * @param suspended    whether or not the argument has its execution suspended
    * @param location     the source location that the node corresponds to
    * @param passData     the pass metadata associated with this node
    * @param diagnostics  compiler diagnostics for this node
    */
  sealed case class Specified(
    override val name: Name,
    override val ascribedType: Option[Expression],
    override val defaultValue: Option[Expression],
    override val suspended: Boolean,
    override val location: Option[IdentifiedLocation],
    override val passData: MetadataStorage      = MetadataStorage(),
    override val diagnostics: DiagnosticStorage = DiagnosticStorage()
  ) extends DefinitionArgument
      with IRKind.Primitive {
    override protected var id: Identifier = randomId

    /** Creates a copy of `this`.
      *
      * @param name         the name of the argument
      * @param ascribedType the explicitly ascribed type of the argument, if
      *                     present
      * @param defaultValue the default value of the argument, if present
      * @param suspended    whether or not the argument has its execution suspended
      * @param location     the source location that the node corresponds to
      * @param passData     the pass metadata associated with this node
      * @param diagnostics  compiler diagnostics for this node
      * @param id           the identifier for the new node
      * @return a copy of `this`, updated with the specified values
      */
    def copy(
      name: Name                           = name,
      ascribedType: Option[Expression]     = ascribedType,
      defaultValue: Option[Expression]     = defaultValue,
      suspended: Boolean                   = suspended,
      location: Option[IdentifiedLocation] = location,
      passData: MetadataStorage            = passData,
      diagnostics: DiagnosticStorage       = diagnostics,
      id: Identifier                       = id
    ): Specified = {
      val res = Specified(
        name,
        ascribedType,
        defaultValue,
        suspended,
        location,
        passData,
        diagnostics
      )
      res.id = id
      res
    }

    override def withName(ir: Name): DefinitionArgument = copy(name = ir)

    /** @inheritdoc */
    override def duplicate(
      keepLocations: Boolean   = true,
      keepMetadata: Boolean    = true,
      keepDiagnostics: Boolean = true,
      keepIdentifiers: Boolean = false
    ): Specified =
      copy(
        name = name.duplicate(
          keepLocations,
          keepMetadata,
          keepDiagnostics,
          keepIdentifiers
        ),
        ascribedType = ascribedType.map(
          _.duplicate(
            keepLocations,
            keepMetadata,
            keepDiagnostics,
            keepIdentifiers
          )
        ),
        defaultValue = defaultValue.map(
          _.duplicate(
            keepLocations,
            keepMetadata,
            keepDiagnostics,
            keepIdentifiers
          )
        ),
        location = if (keepLocations) location else None,
        passData = if (keepMetadata) passData.duplicate else MetadataStorage(),
        diagnostics =
          if (keepDiagnostics) diagnostics.copy else DiagnosticStorage(),
        id = if (keepIdentifiers) id else randomId
      )

    /** @inheritdoc */
    override def setLocation(
      location: Option[IdentifiedLocation]
    ): Specified = copy(location = location)

    /** @inheritdoc */
    def mapExpressions(fn: Expression => Expression): Specified = {
      copy(
        name         = name.mapExpressions(fn),
        ascribedType = ascribedType.map(fn),
        defaultValue = defaultValue.map(fn)
      )
    }

    /** @inheritdoc */
    override def toString: String =
      s"""
         |DefinitionArgument.Specified(
         |name = $name,
         |ascribedType = $ascribedType,
         |defaultValue = $defaultValue,
         |suspended = $suspended,
         |location = $location,
         |passData = ${this.showPassData},
         |diagnostics = $diagnostics,
         |id = $id
         |)
         |""".toSingleLine

    /** @inheritdoc */
    override def children: List[IR] =
      name :: ascribedType.toList ++ defaultValue.toList

    /** @inheritdoc */
    override def showCode(indent: Int): String = {
      val withoutLazy =
        if (defaultValue.isDefined && ascribedType.isDefined) {
          val name        = this.name.showCode(indent)
          val typeExpr    = this.ascribedType.get.showCode(indent)
          val defaultExpr = this.defaultValue.get.showCode(indent)
          s"($name : ($typeExpr) = ($defaultExpr))"
        } else if (defaultValue.isDefined) {
          val name        = this.name.showCode(indent)
          val defaultExpr = this.defaultValue.get.showCode(indent)
          s"($name = $defaultExpr)"
        } else if (ascribedType.isDefined) {
          val name     = this.name.showCode(indent)
          val typeExpr = this.ascribedType.get.showCode(indent)
          s"($name : $typeExpr)"
        } else {
          s"${name.showCode(indent)}"
        }

      if (suspended) {
        s"~$withoutLazy"
      } else {
        withoutLazy
      }
    }
  }
}

File: engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/DefinitionArgument.java
package org.enso.compiler.core.ir;

import java.util.function.Function;
import org.enso.compiler.core.IR;
import org.enso.runtime.parser.dsl.GenerateFields;
import org.enso.runtime.parser.dsl.GenerateIR;
import org.enso.runtime.parser.dsl.IRChild;
import org.enso.runtime.parser.dsl.IRField;
import scala.Option;

public interface DefinitionArgument extends IR {
  Name name();

  Option<Expression> ascribedType();

  Option<Expression> defaultValue();

  boolean suspended();

  @Override
  DefinitionArgument mapExpressions(Function<Expression, Expression> fn);

  @Override
  DefinitionArgument duplicate(
      boolean keepLocations,
      boolean keepMetadata,
      boolean keepDiagnostics,
      boolean keepIdentifiers);

  DefinitionArgument withName(Name ir);

  @GenerateIR(interfaces = {DefinitionArgument.class})
  final class Specified extends DefinitionArgumentSpecifiedGen {
    @GenerateFields
    public Specified(
        @IRChild Name name,
        @IRChild Option<Expression> ascribedType,
        @IRChild Option<Expression> defaultValue,
        @IRField boolean suspended,
        IdentifiedLocation identifiedLocation,
        MetadataStorage passData,
        DiagnosticStorage diagnostics) {
      super(name, ascribedType, defaultValue, suspended, identifiedLocation, passData, diagnostics);
    }

    public static Builder builder() {
      return new Builder().ascribedType(Option.empty()).defaultValue(Option.empty());
    }

    public static Builder builder(Specified original) {
      return new Builder(original);
    }

    @Override
    public String showCode(int indent) {
      String withoutLazy;
      if (defaultValue().isDefined() && ascribedType().isDefined()) {
        var name = name().showCode(indent);
        var typeExpr = ascribedType().get().showCode(indent);
        var defaultExpr = defaultValue().get().showCode(indent);
        withoutLazy = String.format("(%s : (%s) = (%s))", name, typeExpr, defaultExpr);
      } else if (defaultValue().isDefined()) {
        var name = name().showCode(indent);
        var defaultExpr = defaultValue().get().showCode(indent);
        withoutLazy = String.format("(%s = %s)", name, defaultExpr);
      } else if (ascribedType().isDefined()) {
        var name = name().showCode(indent);
        var typeExpr = ascribedType().get().showCode(indent);
        withoutLazy = String.format("((%s : %s))", name, typeExpr);
      } else {
        withoutLazy = name().showCode(indent);
      }
      if (suspended()) {
        return "~" + withoutLazy;
      } else {
        return withoutLazy;
      }
    }

    @Override
    public DefinitionArgument withName(Name ir) {
      return new Builder(this).name(ir).build();
    }

    public Specified copyWithAscribedType(Option<Expression> ascribedType) {
      return new Builder(this).ascribedType(ascribedType).build();
    }

    public Specified copyWithDefaultValue(Option<Expression> defaultValue) {
      return new Builder(this).defaultValue(defaultValue).build();
    }

    public Specified copyWithSuspended(boolean suspended) {
      return new Builder(this).suspended(suspended).build();
    }

    public Specified copy(Option<Expression> defaultValue, Option<Expression> ascribedType) {
      return new Builder(this).defaultValue(defaultValue).ascribedType(ascribedType).build();
    }

    public Specified copy(Name name, Option<Expression> defaultValue) {
      return new Builder(this).name(name).defaultValue(defaultValue).build();
    }

    public Specified copyWithNameAndAscribedType(Name name, Option<Expression> ascribedType) {
      return new Builder(this).name(name).ascribedType(ascribedType).build();
    }
  }
}
